docs/proposals/20200304-autoscaler-scale-ux.md

```diff
@@ -0,0 +1,117 @@
+---
+title: Cluster Autoscaler Scale from Zero UX
+authors:
+  - "@michaelgugino"
+reviewers:
+  - "@detiber"
+creation-date: 2020-03-04
+creation-date: 2020-03-04
+status:
+---
+
+# Cluster Autoscaler Scale from Zero UX
+
+## Summary
+Cluster Autoscaler supports scale from zero for cloud-native providers. This is done via compiling CPU, GPU, and RAM info into the cluster autoscaler at build time for each supported cloud.
+
+We are attempting to add cluster-api support to the cluster-autoscaler to scale machinesets, machinedeployments, and/or other groupings of machines as appropriate. We are trying to keep an abstraction which is isolated from particular cloud implementations. We need a way to inform the cluster-autoscaler of the value of machine resources so it can make proper scheduling decisions when there are not currently any machines in a particular scale set (machineset/machinedeployment, etc).
+
+We will need input from the autoscaler team, as well as input from api reviewers to ensure we obtain the best solution.
+
+These ideas require a particular (possibly new) cluster-api cloud provider controller
+which can inspect machinesets and annotate them as appropriate.  This could be
+done dynamically via billing API (not recommended) or have values compiled in
+statically.
+
+We could also attempt to build this logic into the machineset/machinedeployment
+controller, but this would probably not be very scalable and would require
+a well defined set of fields on the provider CRD.
+
+## Motivation
+
+### Goals
+
+### Non-Goals
+
+## Proposal
+
+### User Stories
+
+As a cluster admin, I want to add the ability to automatically scale my cluster
+using cluster autoscaler and the cluster-api.  In order to save money, I need
+the ability to scale from zero for a given machineset or machinedeployment.
+
+## Proposal Ideas
+
+There are several proposals to consider.  Once we decide on one, we'll eliminate
+the others.
+
+### Annotations
+Modify the autoscaler to look at particular annotations on a given object to determine the appropriate amounts of each resource.
+
+#### Benefits
+Easy, simple.  Users can modify if needed (we can instruct controllers to only
+add annotations if they're missing, etc), users can bring their own clouds
+and annotate their objects to support scale from zero and we don't need to worry
+about supporting every cloud day 1.
+
+#### Cons
+We're essentially creating an API around an annotation, and the api folks don't
+like that type of thing.
```

Comment:
Can we elaborate on the downside of doing that? Not sure "the api folks don't
like that type of thing" counts :)

Reply:
I'm think they don't like to have specific annotations be part of a defined api.  We might be able to do just annotations if this was all for a single controller, but with an external controller looking at our object(s), it's unlikely to pass API review.

Reply:
The main issues with annotations is that they are an informal api that is detached from the actual api, so enforcing forward/backward compatibility, migrations for support, etc are quite difficult.

For example, the cluster-autoscaler currently uses the deprecated `cluster.k8s.io/delete-machine` annotation rather than `cluster.x-k8s.io/delete-machine` and in order to not break the integration we need to handle the backwards/forwards compatibility both in Cluster API and also in cluster-autoscaler: https://github.com/kubernetes/autoscaler/pull/3161.

If the integration was done using an API field, then conversion webhooks would help ease the migration.